stataggregator/countaggregator.go

```diff
@@ -0,0 +1,115 @@
+package stataggregator
+
+import (
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rs/xstats"
+)
+
+// CountAggregator is a wrapper around xstats.XStater, that aggregates
+// Count metrics on a time interval before sending them through Stater
+type CountAggregator struct {
+	Stater        xstats.XStater
+	Bucket        map[StatTagKey]float64
+	lock          *sync.Mutex
```

Comment:
Interesting suggestion. I'm not sure if it would.. My understanding of RWMutex is that there is a single lock, and locking around reading doesn't block other reads, but blocks writing. Locking around writing blocks both writing and reading. Most of these operations involve writing, but I suppose you could have more fine-grained locking around the actual code parts that read and write. I'm not sure how much extra benefit is in this, fine-grained locking eventually has diminishing returns.

But, on the topic of fine-grained locking, suppose we introduce locks around each entry in the `Bucket`. This could probably lead to better control, which allows users to concurrently emit different stats with different tags that don't block each other. This would be kinda tricky to implement, but stream of consciousness 🤷

Reply:
I think we can take it as an improvement afterwards, considering this functionality is micro batching metrics as traffic on API increases micro batching could start consuming more CPU. A perf tests can revel such bottlenecks. Additional patter to separate core functionality with metrics could be to generate metrics from logs instead. It seems like we do have multiple paths forward, do we track such follow ups as part of ticket?

Reply:
will make a ticket!